libs/filer/workspace_files_cache.go

```diff
@@ -0,0 +1,425 @@
+func (c *cache) Cleanup() {
+	close(c.queue)
+	c.wg.Wait()
+}
+
+// Write passes through to the underlying Filer.
+func (c *cache) Write(ctx context.Context, name string, reader io.Reader, mode ...WriteMode) error {
+	return c.f.Write(ctx, name, reader, mode...)
+}
+
+// Read passes through to the underlying Filer.
+func (c *cache) Read(ctx context.Context, name string) (io.ReadCloser, error) {
+	return c.f.Read(ctx, name)
+}
+
+// Delete passes through to the underlying Filer.
+func (c *cache) Delete(ctx context.Context, name string, mode ...DeleteMode) error {
+	return c.f.Delete(ctx, name, mode...)
+}
+
+// Mkdir passes through to the underlying Filer.
+func (c *cache) Mkdir(ctx context.Context, name string) error {
+	return c.f.Mkdir(ctx, name)
+}
+
+// ReadDir returns the entries in a directory.
+// If the directory is already in the cache, the cached value is returned.
+func (c *cache) ReadDir(ctx context.Context, name string) ([]fs.DirEntry, error) {
+	name = path.Clean(name)
+
+	// Lock before R/W access to the cache.
+	c.m.Lock()
+
+	// If the directory is already in the cache, wait for and return the cached value.
+	if e, ok := c.readDir[name]; ok {
+		c.m.Unlock()
+		return e.wait(ctx)
+	}
+
+	// Otherwise, create a new cache entry and queue the operation.
+	e := newReadDirEntry(name)
+	err := c.enqueue(ctx, e)
+	if err != nil {
+		c.m.Unlock()
+		return nil, err
+	}
+
+	c.readDir[name] = e
+	c.m.Unlock()
+
+	// Wait for the operation to complete.
+	return e.wait(ctx)
+}
+
+// statFromReadDir returns the file info for a file or directory.
+// If the file info is already in the cache, the cached value is returned.
+func (c *cache) statFromReadDir(ctx context.Context, name string, entry *readDirEntry) (fs.FileInfo, error) {
+	_, err := entry.wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Upon completion of a [ReadDir] call, all directory entries are added to the stat cache and
+	// enqueue a [Stat] call if necessary (entries for notebooks are incomplete and require a
+	// real stat call).
+	//
+	// This means that the file or directory we're trying to stat, either
+	//
+	//   - is present in the stat cache
+	//   - doesn't exist.
+	//
+	c.m.Lock()
+	e, ok := c.stat[name]
+	c.m.Unlock()
+	if ok {
+		return e.wait(ctx)
+	}
+
+	return nil, FileDoesNotExistError{name}
+}
+
+// Stat returns the file info for a file or directory.
+// If the file info is already in the cache, the cached value is returned.
+func (c *cache) Stat(ctx context.Context, name string) (fs.FileInfo, error) {
+	name = path.Clean(name)
+
+	// Lock before R/W access to the cache.
+	c.m.Lock()
+
+	// If the file info is already in the cache, wait for and return the cached value.
+	if e, ok := c.stat[name]; ok {
+		c.m.Unlock()
+		return e.wait(ctx)
+	}
+
+	// If the parent directory is in the cache (or queued to be read),
+	// wait for it to complete to avoid redundant stat calls.
+	dir := path.Dir(name)
```

Comment:
Do you know if this is the primary reason for the performance gain when using this cache? That is we avoid the redundant stat API calls?

Reply:
Not primary, but it helps. Primary I would say is caching in the first place. The extras are a cherry on top.